tensorboard/data/provider.py

```diff
@@ -392,6 +437,110 @@ def __repr__(self):
     ))
 
 
+class TensorTimeSeries(_TimeSeries):
+  """Metadata about a tensor time series for a particular run and tag.
+
+  Attributes:
+    max_step: The largest step value of any datum in this tensor time series; a
+      nonnegative integer.
+    max_wall_time: The largest wall time of any datum in this time series, as
+      `float` seconds since epoch.
+    plugin_content: A bytestring of arbitrary plugin-specific metadata for this
+      time series, as provided to `tf.summary.write` in the
+      `plugin_data.content` field of the `metadata` argument.
+    description: An optional long-form Markdown description, as a `str` that is
+      empty if no description was specified.
+    display_name: An optional long-form Markdown description, as a `str` that is
+      empty if no description was specified. Deprecated; may be removed soon.
+  """
+
+  def __eq__(self, other):
+    if not isinstance(other, TensorTimeSeries):
+      return False
+    if self._max_step != other._max_step:
+      return False
+    if self._max_wall_time != other._max_wall_time:
+      return False
+    if self._plugin_content != other._plugin_content:
+      return False
+    if self._description != other._description:
+      return False
+    if self._display_name != other._display_name:
+      return False
+    return True
+
+  def __hash__(self):
+    return hash((
+        self._max_step,
+        self._max_wall_time,
+        self._plugin_content,
+        self._description,
+        self._display_name,
+    ))
+
+  def __repr__(self):
+    return "TensorTimeSeries(%s)" % ", ".join((
+        "max_step=%r" % (self._max_step,),
+        "max_wall_time=%r" % (self._max_wall_time,),
+        "plugin_content=%r" % (self._plugin_content,),
+        "description=%r" % (self._description,),
+        "display_name=%r" % (self._display_name,),
+    ))
+
+
+class TensorDatum(object):
+  """A single datum in a tensor time series for a run and tag.
+
+  Attributes:
+    step: The global step at which this datum occurred; an integer. This
+      is a unique key among data of this time series.
+    wall_time: The real-world time at which this datum occurred, as
+      `float` seconds since epoch.
+    numpy: The `numpy.ndarray` value with the tensor contents of this
```

Comment:
Asking because I lack knowledge: is a Tensor always ndarray?

Reply:
That’s up to us! :-)

Every “normal” tensor (I’m excluding, e.g., resource tensors) *can* be
represented as an ndarray, via `.numpy()` at eager execution time—

```python
>>> import tensorflow as tf
>>> tf.constant(17).numpy()
17
>>> type(tf.constant(17).numpy())
<class 'numpy.int32'>
>>> tf.constant([42.0, 43.0]).numpy()
array([42., 43.], dtype=float32)
>>> tf.constant(["hello", "world"]).numpy()
array([b'hello', b'world'], dtype=object)
```

—or from proto via `tf.make_ndarray` or our own notf equivalent in
`util/tensor_util.py`. So the choice that I was considering was whether
to return an `ndarray` here or a raw `TensorProto`. I opted to use an
`ndarray` because (a) I think that that’s what most clients will find
much more convenient to use, and (b) it reduces API-level coupling to
TensorFlow’s protos, which TensorFlow appreciates.

The downside is decoding performance for plugins that really want to
operate on the serialized form without a roundtrip. If this becomes a
real use case, we can imagine adding a flag to return the raw protos.

Reply:
> ... I opted to use an ndarray because (a) I think that that’s what most clients will find
much more convenient to use, and (b) it reduces API-level coupling to
TensorFlow’s protos, which TensorFlow appreciates.

Yup, the upside was clear just reading the code.

I actually did not know `np.ndarray(1).shape` would be `()`. Given that a single value can be correctly represented by ndarray with the right shape and right dtype, I now see the benefit of using ndarray here.

Reply:
> I actually did not know `np.ndarray(1).shape` would be `()`.

Yeah, I had to double-check, myself. It’s extra non-obvious because
numpy uses the same `repr` for the rank-0 ndarray as for its item,
unlike TensorFlow’s eager tensors, which are quite clear:

```python
>>> tf.constant(17)
<tf.Tensor: shape=(), dtype=int32, numpy=17>
```

> Given that a single value can be correctly represented by ndarray with
> the right shape and right dtype, I now see the benefit of using
> ndarray here.

Cool; thanks for your input! Glad that we agree.